### PR TITLE
Make SDSS download field names and be more consistent about DRs and constants

### LINKED ISSUE #543: add DR argument for sdss queries
There are significant differences between the DRs of SDSS, it may be useful to have an optional argument which release should be used.
(I guess this is pretty straightforward to implement, one only needs to check that the API remained unchanged between releases and put all the URL in a dictionary, etc.)

### PR DESCRIPTION
**Pull request by eteq**
This PR makes a bunch of (primarily internal) changes that fix a wide range of issues I encountered with the `SDSS` module.  It's partly building from the issues in #469 and #508, it closes #543, and I *think* we can consider it as closing #527 (although @bsipocz might want to comment there if she thinks otherwise).

The main changes are:
* Instead of having a fixed list of field names, the code now downloads the relevant list of columns as part of the query process.  For DR12 (the default), these are built-in so that you don't have to wait for that download if you're using the defaults.
* there are now either ``dr`` or ``drorurl`` arguments to the various query methods.  The idea is that you actually get *different* answers (often very non-trivial differences) that have major effects on the science if you use different data releases.  So users should have to be explicit about this if they aren't using the default (DR12, the last SDSS III release).
* Significant changes to the constants that were previously defined at the class level for `SDSS` but used the configuration object in a way that I think was rather confusing (i.e., users might think they're changing it when actually not).  The current scheme is now more consistent in how it uses all of these things at runtime vs. class creation.  It's also better about using ``self`` everywhere, allowing users to modify settings via attributes of the instance rather than the class itself.

I should probably go through and clean up docs and tests a bit, more, but I thought I would put this up now in case some of it ends up getting changes.

I also noted something very strange that I could add a fix to here, or in a separate PR: the code that searches regions on the sky, which uses the ``radius`` keyword, seems to be wrong: it doesn't search in a circle on the sky (which is what ``radius`` implies), but rather a box with sides ``radius``.  Moreover, it's not even a box in *physical* size, because the cos(dec) factor is missing from the delta-ra.  Is this intended?  Should I change it to instead search a circle (that's my preference based on the word ``radius``), or a box of size ``radius`` in dec and ``radius/cos(dec)`` in ra?

cc @keflavich @bsipocz , and I'm not sure who else has done a lot of the development here

### CODE REVIEW

File: astroquery/sdss/core.py [hunk 1]
```diff
@@ -941,4 +999,10 @@ def _args_to_payload(self, coordinates=None, radius=2. * u.arcsec,
 
         return request_payload
 
+    def _get_query_url(self, drorurl, suffix):
+        if isinstance(drorurl, basestring) and len(drorurl) > 2:
```
Comment (keflavich) [inline astroquery/sdss/core.py]: `basestring` should be `six.string_types`, I believe (for py3 compat)

**Comment by keflavich** [comment]
@eteq - change to circle, and add a box option (width + height kwds) would be my preference for consistency with others that support such queries (SIMBAD).

SDSS development has been by many folks.  @mirochaj did some original work but I'm not sure he's interested any more.  @loicseguin has done some work here.  If either of you want to review this, it would be helpful.

**Comment by keflavich** [comment]
Also, @eteq, if and when you think this is ready to merge, please run the remote tests and report on them (we don't let CI servers run those)

File: astroquery/sdss/core.py (hunk 1 above)
Comment (eteq) [inline astroquery/sdss/core.py]: oops, good point - presumably that'll fix the py3 failures in the tests.

**Comment by eteq** [comment]
@keflavich - re: box: can you point me to where this is done in SIMBAD? (just so I can be sure I do it the exact same way)  I see radius but am not finding width/height.

**Comment by keflavich** [comment]
@eteq - sorry, it was [vizier](https://github.com/astropy/astroquery/blob/master/astroquery/vizier/core.py#L313), not SIMBAD.

**Comment by eteq** [comment]
Aha, got it @keflavich - can do, thanks.

**Comment by bsipocz** [comment]
@eteq - Thanks for doing this and I'm very glad to see the ``dr`` keyword and that you gave it a long overdue cleanup.

**Comment by bsipocz** [comment]
re: radius vs box - i don't have strong feelings which one to use (but if we go with the box we it would be nice to fix the current naive box search to have a physical size on both axis).

**Comment by parejkoj** [comment]
It looks like @eteq's changes were never merged back to upstream. I'll try to merge eteq's fork into mine when working on it, and improve from there.

**Comment by bsipocz** [comment]
@parejkoj - Easiert is to start a branch based on @eteq's one. E.g. you add his for as a remote, and create a branch from that: 

```
git remote add eteq git://github.com/eteq/astroquery
git checkout -b your_branchname eteq/improve_sdss
```

**Comment by parejkoj** [comment]
Oh, you're back! Can I work from his branch directly?

**Comment by bsipocz** [comment]
Yep, just got home, but still has to have some dinner before getting too grumpy. You could work towards this branch as well, with opening PR to it, but I'm not sure whether anyone but @eteq could merge them into this one. 
So I would suggest to start a new one on top of this (with the commands above you basically copy it, but without having any superfluous merge commits), as you have to commit your changes somewhere eventually and can't do it in his remote one.

**Comment by parejkoj** [comment]
I'm working in my fork, same branch. Worst case, we can sort it out later.

**Comment by parejkoj** [comment]
It looks like test_sdss doesn't actually test anything? There's almost no asserts in the test functions.

**Comment by bsipocz** [comment]
Well, there is room for improvement ;) There are more functional tests in ``test_sdss_remote``.

**Comment by bsipocz** [comment]
@parejkoj - I suggest to move the chat to gitter (https://gitter.im/astropy/astroquery), to avoid generating too much off topic discussion under this PR.

**Comment by parejkoj** [comment]
I don't like the "drorurl" thing: can we tear that out and replace it with just "dr="? Replacing the entirety of the url seems sketchy to me.

**Comment by bsipocz** [comment]
I'm not sure about the ``drorurl`` stuff either, maybe @eteq has a use case in mind when it's useful to be able to pass an url, rather than figuring it out from a built-in list based on the dr number.

**Comment by parejkoj** [comment]
In that case, the url should be a different parameter: overloading that for either int or string is bad.

**Comment by bsipocz** [comment]
Good point. In that case ``url=None`` by default, and if given overrides ``dr``?

**Comment by parejkoj** [comment]
That's what I was thinking. But I'd just not include url unless given a reason to do so.

**Comment by eteq** [comment]
Alright, I've updated this with contributions from @parejkoj @cmp346 @adamamiller, and @weaverba137 .  I also ran the remote-data tests and they all passed.

So if the travis tests pass, I'll go ahead and merge this so that everyone can be working from a more up-to-date master.

**Comment by eteq** [comment]
Also, the plan is to still address the circle matching issue, but in a separate PR.  Ditto with some other related fixes that are being discussed on gitter.

@bsipocz and @parejkoj - to clarify the `drorurl` business: My motivation for that is as follows: it's intended to support the possibility of future DRs that don't follow the previous patterns.  As you of course know, DR12 has a URL that's different in more ways than just the number "12", and I found it very hacky to support that.  So this is supposed to allow future users to say "ah, screw it, I'm just giving the URL for the query directly!" without having to dig into the plumbing of the code too much.

**Comment by bsipocz** [comment]
@eteq - There are still many merge commits in here, including stuff from unrelated modules.

**Comment by parejkoj** [comment]
@eteq that's fine, but specifying the url should absolutely be a different argument. And I'd argue that the correct thing is to just update the code to support any new SDSS url (because the access api may change as well).

I'll change things to just use dr=, get rid of drurl=, and we can add url= (overrides dr=) later.

**Comment by eteq** [comment]
@bsipocz - the merge commits are OK because there's no easy way to get rid of them at this point... But you're right there are a few unrelated things... Will try to rebase to expunge those

**Comment by eteq** [comment]
@parejkoj - that all sounds fine to me (re: `drurl`->`dr` and `url`)

**Comment by eteq** [comment]
I tested this locally (with and without remote, with 2.x and 3.x) and its now all working after various cherry-picking of the other commits.  So to let others continuing working on the hack day I'm merging.  But I sure hope it doesn't kill travis :(

**Comment by bsipocz** [comment]
Merged, so others can start from the top of it to fix the remaining issues.

**Pull request merged**